src/hooks/use-dropdown-menu.tsx

```diff
@@ -0,0 +1,153 @@
+// Imports
+import React, { useState, useRef, createRef, useEffect } from 'react';
+
+
+// A custom Hook that abstracts away the listeners/controls for dropdown menus
+export default (itemCount: number) => {
+   // Use state
+   const [isOpen, setIsOpen] = useState<boolean>(false);
+   const [returnFocusToButton, setReturnFocusToButton] = useState<boolean>(false);
+   const currentFocusIndex = useRef<number | null>(null);
+   const firstRun = useRef(true);
+   
+   
+   // Create refs
+   const buttonRef = useRef<HTMLButtonElement>(null);
+   const itemRefs = useRef([...Array(itemCount)].map(() => createRef<HTMLAnchorElement>()));
+   
+   
+   // Create type guard
+   const isKeyboardEvent = (e: React.KeyboardEvent | React.MouseEvent): e is React.KeyboardEvent => (e as React.KeyboardEvent).key !== undefined;
+   
+   
+   // Handle changes in the state of the menu
+   useEffect(() => {
+       // Stop if the ref currently points to a null value
+       if (!itemRefs.current[0].current || !buttonRef.current) {
+           return;
+       }
+
+
+       // Stop if this is the first fire of the Hook, and update the ref
+       if (firstRun.current) {
+           firstRun.current = false;
+           return;
+       }
+
+       
+       // If the menu is currently open focus on the first item in the menu, else return focus to the button if state indicates that is desired behavior
+       if (isOpen) {
+           itemRefs.current[0].current.focus();
+           moveFocus(0);
+       } else if (returnFocusToButton) {
+           buttonRef.current.focus();
+       }
+   }, [isOpen, returnFocusToButton]);
+
+
+   const moveFocus = (itemIndex: number) => {
+        currentFocusIndex.current = itemIndex
+
+
+        if (currentFocusIndex?.current) {
```

Comment:
This doesn't seem like the correct logic; first, you're always setting `currentFocusIndex.current = itemIndex`. And we know via TypeScript that `itemIndex` is a number. Which means this conditional will always be true, unless the focus index is `0`. But sometimes we'll want to move the focus to index `0`.

Reply:
Perhaps,
```js
if (currentFocusIndex.current !== null) {
...
}
```

would be a better condition, the compiler complained before that the currentFocusIndex could possibly be null because that is the value it initializes at, so I think it will still need some check.

Reply:
I don't think it needs the conditional at all; TypeScript should realize whenever you reassign it above that it's now a `number` type.